### PR TITLE
Multi step workflow

### PR DESCRIPTION
**Pull request by antho1404**
Fix https://github.com/mesg-foundation/engine/issues/1197

TODO:

- [x] Add multiple tasks to workflows
- [x] Create hash for workflows
- [x] Add workflow hash to execution
- [x] Resolve next steps on workflow based on workflowHash + execution height (length of the chain)

## Testing:

Note that the CLI is not yet updated for that so you will need to use the GRPC API directly.

We will create the following workflow:

```
+------------+      +------------+      +------------+     +------------+
|  ServiceA  |      |  ServiceB  |      |  ServiceC  |     |  ServiceB  |
|   taskX    +----->+   taskY    +----->+   taskZ    +---->+   taskY    |
+------------+      +------------+      +------------+     +------------+
```
```
when Result of ServiceA#taskX
  then Call ServiceB#taskY
  then Call ServiceC#taskZ
  then Call ServiceB#taskY
```

These services are just accumulator that take a message in input and return a message that is the concatenation of the input plus the task executed.

You can find the services here:

**ServiceA**:
```json
{"name":"serviceA","tasks":[{"inputs":[{"type":"String","key":"message"}],"outputs":[{"type":"String","key":"message"}],"key":"taskX"},{"key":"emits"}],"events":[{"data":[{"type":"String","key":"message"}],"key":"eventX"}],"configuration":{},"hash":"9P49TPmeToJDb2vSLwv3iovz56UWJJinRMhM3TMkEcEw","sid":"serviceA","source":"QmUojyMjo2JUvhktLU4royUBDEsVzU4z1onPovGpVQAFDW"}
```

**ServiceB**:
```json
{"name":"serviceB","tasks":[{"inputs":[{"type":"String","key":"message"}],"outputs":[{"type":"String","key":"message"}],"key":"taskY"}],"configuration":{},"hash":"3vZaZD59AUzxMAaRREFFR2XEKvU6mYDVv5UweTgSQF4c","sid":"serviceB","source":"Qmf2e92jNxxmYyjxEFrCBJdCNUS1LKNWet2ZQt3ByEUpAP"}
```

**ServiceC**:
```json
{"name":"serviceC","tasks":[{"inputs":[{"type":"String","key":"message"}],"outputs":[{"type":"String","key":"message"}],"key":"taskZ"},{"inputs":[{"type":"String","key":"message"}],"outputs":[{"type":"String","key":"message"}],"key":"taskA"}],"configuration":{},"hash":"EvYgwFjAQpGs19EgoXi4EExJXoH7DL6Fmae4RStQZzt2","sid":"serviceC","source":"QmPuHX3LumnDkrxY3wUR8J2J9xmXGw8yTVy5jf7P24MprN"}
```

Once these services are created and started you can write the following workflow:
```json
{
  "sid": "workflow",
  "name": "workflow",
  "source": "QmbtKRgiDxB8dChfdRA2Hk7MYtB1yB3dXrnh2msus7Ft8P",
  "workflows": [
    {
      "key": "event",
      "trigger": {
        "type": 1,
        "instanceHash": "88YtHw3iZtjpAeDDUWHsDFg9MLvgircPzCFHxFqJ2ex9",
        "key": "eventX"
      },
      "tasks": [
        {
          "instanceHash": "8qKUdaMPAn35ozeUWVcT9BQ5gHNJYbJJ5rTCWqfnJgpW",
          "taskKey": "taskY"
        },
        {
          "instanceHash": "7sTRCDpWCXYBSVgh7dKKV6pP69PHdYQY8kyBohumEWNu",
          "taskKey": "taskZ"
        }
      ]
    },
    {
      "key": "result",
      "trigger": {
        "type": 2,
        "instanceHash": "88YtHw3iZtjpAeDDUWHsDFg9MLvgircPzCFHxFqJ2ex9",
        "key": "taskX"
      },
      "tasks": [
        {
          "instanceHash": "8qKUdaMPAn35ozeUWVcT9BQ5gHNJYbJJ5rTCWqfnJgpW",
          "taskKey": "taskY"
        },
        {
          "instanceHash": "7sTRCDpWCXYBSVgh7dKKV6pP69PHdYQY8kyBohumEWNu",
          "taskKey": "taskZ"
        }
      ]
    },
    {
      "key": "result-from-other-workflow",
      "trigger": {
        "type": 2,
        "instanceHash": "8qKUdaMPAn35ozeUWVcT9BQ5gHNJYbJJ5rTCWqfnJgpW",
        "key": "taskY"
      },
      "tasks": [
        {
          "instanceHash": "7sTRCDpWCXYBSVgh7dKKV6pP69PHdYQY8kyBohumEWNu",
          "taskKey": "taskZ"
        }
      ]
    }
  ]
}
```
Where
- `88YtHw3iZtjpAeDDUWHsDFg9MLvgircPzCFHxFqJ2ex9`: Hash of instance of serviceA
- `8qKUdaMPAn35ozeUWVcT9BQ5gHNJYbJJ5rTCWqfnJgpW`: Hash of instance of serviceB
- `7sTRCDpWCXYBSVgh7dKKV6pP69PHdYQY8kyBohumEWNu`: Hash of instance of serviceC

Now if you execute **taskX** from **serviceA** all the tasks of the workflow will be triggered.

```
grpcurl -plaintext \
  -d "{\"instanceHash\":\"88YtHw3iZtjpAeDDUWHsDFg9MLvgircPzCFHxFqJ2ex9\", \"taskKey\":\"taskX\", \"inputs\":{\"message\":\"\"}}" \
  localhost:50052 api.Execution/Create
```

You can also try with events by calling the `emits` task from `serviceA` with no inputs
```
grpcurl -plaintext \
  -d "{\"instanceHash\":\"88YtHw3iZtjpAeDDUWHsDFg9MLvgircPzCFHxFqJ2ex9\", \"taskKey\":\"emits\", \"inputs\":{}}" \
  localhost:50052 api.Execution/Create
```

### CODE REVIEW

File: workflow/workflow.go [hunk 1]
```diff
@@ -62,12 +64,59 @@ func (w *Workflow) process(trigger service.TriggerType, instanceHash hash.Hash,
...
 					w.ErrC <- err
-					continue
 				}
 			}
 		}
 	}
 }
+
+func (w *Workflow) processExecution(exec *execution.Execution) error {
+	if exec.WorkflowHash.IsZero() {
```
Comment (krhubert) [inline workflow/workflow.go]: What is the current state of workflow schema? Is workflow part of service or not? Why core needs workflow to execute single execution?

File: workflow/workflow.go [hunk 2]
```diff
@@ -62,12 +64,59 @@ func (w *Workflow) process(trigger service.TriggerType, instanceHash hash.Hash,
...
+	if len(wf.Tasks) <= height {
+		// end of workflow
+		return nil
+	}
+	task := wf.Tasks[height]
+	hash, err := w.execution.Execute(wf.Hash, task.InstanceHash, eventHash, prev.Hash, task.TaskKey, data, []string{})
+	if err != nil {
+		return err
+	}
+	logrus.WithFields(logrus.Fields{
```
Comment (krhubert) [inline workflow/workflow.go]: remove logs

File: workflow/workflow.go [hunk 3]
```diff
@@ -62,12 +64,59 @@ func (w *Workflow) process(trigger service.TriggerType, instanceHash hash.Hash,
...
+		return 0, nil
+	}
+	if exec.ParentHash == nil {
+		return 0, nil
+	}
+	parent, err := w.execution.Get(exec.ParentHash)
+	if err != nil {
+		return 0, err
+	}
+	parentHeight, err := w.getHeight(parent)
```
Comment (krhubert) [inline workflow/workflow.go]: possible stackoverflow if hash == parenthash. It shoudn't happen at all but we can add this check, what do you think?

File: protobuf/types/service.proto [hunk 4]
```diff
@@ -94,9 +94,10 @@ message Service {
       string taskKey = 2;       // Task of the instance to execute.
     }
 
-    string key = 3;       // Workflow's key
-    Trigger trigger = 1;  // Trigger for the workflow.
-    Task task = 2;        // Task to execute when the trigger is valid.
+    string hash = 4;          // Workflow's hash
```
[the remark this answers is not in the document]
Comment (antho1404) [reply protobuf/types/service.proto]: service has a hash with 10, this doesn't matter much. I agree one day we will need to cleanup the whole api but nothing critical at all

File: workflow/workflow.go (hunk 3 above)
Comment (antho1404) [reply workflow/workflow.go]: just put a panic in that case.

File: workflow/workflow.go (hunk 1 above)
Comment (antho1404) [reply workflow/workflow.go]: Workflow is still part of the service but will certainly move into its own ressource and api soon.

> Why core needs workflow to execute single execution

That's the whole purpose of workflows to automatically trigger executions based on some definition. I'm not sure I really understand your question here

File: execution/execution.go [hunk 5]
```diff
@@ -35,6 +35,7 @@ func (s Status) String() (r string) {
 // Execution stores all information about executions.
 type Execution struct {
 	Hash         hash.Hash              `hash:"-"`
+	WorkflowHash hash.Hash              `hash:"name:workflowHash"`
```
[the remark this answers is not in the document]
Comment (antho1404) [reply execution/execution.go]: If you don't have the link to the workflow that triggers this execution how can you continue the chain of execution?
One execution belongs to one workflow only but one workflow can trigger multiple executions and multiple workflows can trigger a similar execution (same event that triggers the same task but with a different context)

I don't know if that answers your questions

**Comment by antho1404** [comment]
Let's merge this pull request and continue the work on another PR.

**Pull request merged**